.github/actions/setup-runner/action.yml

```diff
@@ -0,0 +1,84 @@
+name: Setup runner
+description: Setup the runner with the necessary tools and dependencies.
+
+inputs:
+  flutter-version:
+    description: Flutter version to use.
+    default: stable
+  cache-key:
+    description: Cache key to use for restoring and saving the cache.
+    default: ${{ hashFiles('**/pubspec.yaml') }}
+
+
+runs:
+  using: "composite"
+  steps:
+    - name: Setup environment
+      # By forcing puro (and hence the pub-cache) onto the same drive as the checkout,
+      # we can avoid symlink issue between local and network drives on windows.
+      shell: bash
+      run: |
+        export PURO_ROOT=$(dirname $GITHUB_WORKSPACE)/puro
+        mkdir -p $PURO_ROOT
+
+        echo $PURO_ROOT/bin >> $GITHUB_PATH
+        echo $PURO_ROOT/shared/pub_cache/bin/ >> $GITHUB_PATH
+        echo $PURO_ROOT/envs/default/flutter/bin/ >> $GITHUB_PATH
+
+        echo PURO_ROOT=$PURO_ROOT >> $GITHUB_ENV
+        echo PUB_CACHE=$PURO_ROOT/shared/pub_cache >> $GITHUB_ENV
+        echo MELOS_SDK_PATH=$PURO_ROOT/envs/default/flutter/ >> $GITHUB_ENV
+
+    - name: Restore cache
+      # TODO: Avoid skipping restore on Windows
+      ###if: runner.os != 'Windows'
+      id: restore
+      uses: actions/cache/restore@v4
+      with:
+        path: ${{ env.PURO_ROOT }}
+        key: setup-runner-${{ runner.os }}-${{ runner.arch }}-${{ inputs.cache-key }}
+        restore-keys: setup-runner-${{ runner.os }}-${{ runner.arch }}-
+
+    - name: Install puro (posix)
+      # See https://puro.dev/
+      # Skip on any cache hit (partial or full)
+      if: runner.os != 'Windows' && steps.restore.outputs.cache-matched-key == ''
+      shell: bash
+      run: curl -sS -o- https://puro.dev/install.sh | PURO_VERSION="1.4.6" bash
+
+    - name: Install puro (windows)
+      # We always install on windows, as we currently don't restore a cache (see TODO above)
+      if: runner.os == 'Windows' && steps.restore.outputs.cache-matched-key == ''
+      shell: pwsh
+      run: >-
+        Invoke-WebRequest
+        -Uri "https://puro.dev/builds/1.4.6/windows-x64/puro.exe"
+        -OutFile "$Env:PURO_ROOT";
+        &"$Env:PURO_ROOT\puro.exe" install-puro --promote
+
+    - name: Upgrade puro and flutter
+      shell: bash
+      run: |
+        puro upgrade-puro
+        puro use --global ${{ inputs.flutter-version }}
+        puro upgrade ${{ inputs.flutter-version }}
+        puro gc
+
+    - name: Activate melos
+      if: steps.restore.outputs.cache-matched-key == ''
+      shell: bash
+      run: puro pub global run melos --version || puro pub global activate melos
+
+    - name: Melos bootstrap
+      shell: bash
+      run: puro pub global run melos bootstrap
+
+    - name: Save cache
+      # Saving on feature branches will only benefit reruns, so we don't bother.
+      # Feature branches inherit the cache from the main branch anyway.
+      # This saves on cache storage.
+      if: github.ref != 'refs/heads/main' && steps.restore.outputs.cache-hit != 'true'
```

Comment:
Should this be `github.ref == 'refs/heads/main'`? Also, we seem to be only be storing the cache if we don't have a cache hit, but then we use the cache primary key as input to the save action, which seems odd.

Reply:
@nirinchev You are a bit early 😊 . Yes it should be `==`.. the `!=` is just for debugging.

You cannot update a cache in github actions, only write a new one, so if you have an exact hit, trying save on the same key would fail. The primary key is the key we tried to lookup, and where we will store the updated cache, if either none existed, or the one matched pertained to a stale set of `pubspec.yaml` files.